invenio_records_permissions/generators.py

```diff
@@ -6,227 +6,192 @@
 from invenio_access.permissions import any_user
 from invenio_files_rest.models import Bucket, ObjectVersion
 from invenio_records_files.api import Record
 from invenio_records_files.models import RecordsBuckets
 
 
-class _NeedClass(object):
-
-    def needs(self):
-        pass
-
-    def excludes(self):
-        pass
+class Generator(object):
+    """Parent class mapping the context when an action is allowed or denied.
 
-    def query_filter(self):
-        pass
+    It does so by *generating* "needed" and "excluded" Needs. At the search
+    level it implements the *query filters* to restrict the search.
 
+    Any context inherits from this class.
+    """
 
-class _AnyUser(_NeedClass):
+    def needs(self, **kwargs):
+        """Enabling low-level Needs."""
+        return []
 
-    def __init__(self):
-        super(_AnyUser, self).__init__()
-
-    def needs(self):
-        return [any_user]
+    def excludes(self, **kwargs):
+        """Preventing low-level Needs."""
+        return []
 
-    def query_filter(self):
-        return Q('match_all')  # match all query
+    def query_filter(self, **kwargs):
+        """Elasticsearch filters."""
+        return []
 
 
-AnyUser = _AnyUser()
-
-
-class _Deny(_NeedClass):
+class AnyUser(Generator):
+    """Allows any user."""
 
     def __init__(self):
-        super(_Deny, self).__init__()
+        """Constructor."""
+        super(AnyUser, self).__init__()
 
-    def excludes(self):
+    def needs(self, **kwargs):
+        """Enabling low-level Needs."""
         return [any_user]
 
-    def query_filter(self):
-        return ~Q('match_all')  # Match None
-
+    def query_filter(self, **kwargs):
+        """Match all in search."""
+        return Q('match_all')
 
-Deny = _Deny()
 
-
-class _Admin(_NeedClass):
+class Deny(Generator):
+    """Denies ALL users (except super users)."""
 
     def __init__(self):
-        super(_Admin, self).__init__()
-
-    def needs(self):
-        return [ActionNeed('admin-access')]
+        """Constructor."""
+        super(Deny, self).__init__()
 
+    def excludes(self, **kwargs):
+        """Preventing low-level Needs."""
+        return [any_user]
 
-Admin = _Admin()
+    def query_filter(self, **kwargs):
+        """Match None in search."""
+        return ~Q('match_all')
 
 
-class _RecordNeedClass(_NeedClass):
+class Admin(Generator):
+    """Allows users with admin-access (different from superuser-access)."""
 
     def __init__(self):
-        super(_RecordNeedClass, self).__init__()
-
-    def needs(self, record):
-        pass
-
-    def excludes(self, record):
-        pass
+        """Constructor."""
+        super(Admin, self).__init__()
 
+    def needs(self, **kwargs):
+        """Enabling low-level Needs."""
+        return [ActionNeed('admin-access')]
 
-class _RecordOwners(_RecordNeedClass):
 
-    def __init__(self):
-        super(_RecordOwners, self).__init__()
+class RecordOwners(Generator):
+    """Allows record owners."""
 
```

Comment:
Removing the constructor call to `super` would make this one not have the `exclude` initialize and would fail? Not sure about this behaviour. To be tested. However, for consistency and safety, shouldn't we have a constructor as for the others?